support multi conditions in apicall

KunWuLuan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

As a kube-scheduler plugin developer, I have to update Pod conditions in my plugin. When multi conditions have to be updated, old api calls will be overwritten by the new api calls, so I want to update multi conditions in one API call.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
#135159

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Changed the `PatchPodStatus` API in the scheduler framework to accept a slice of Pod conditions (`[]*v1.PodCondition`) instead of a single condition (`*v1.PodCondition`). This allows scheduler plugins to update multiple Pod conditions in a single API call, preventing newer calls from overwriting older ones when multiple conditions need to be updated concurrently.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

KunWuLuan committed:
support multi conditions in apicall

Signed-off-by: KunWuLuan <kunwuluan@gmail.com>

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -45,17 +46,17 @@ type PodStatusPatchCall struct {
 	// podStatus contains the actual status of the pod.
 	podStatus *v1.PodStatus
 	// newCondition is a condition to update.
```
```suggestion
	// newConditions is a list of conditions to update.
```

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -88,18 +89,29 @@ func (psuc *PodStatusPatchCall) Execute(ctx context.Context, client clientset.In
 	// Executed flag is set not to race with podStatus write in Sync afterwards.
 	psuc.executed = true
-	condition := psuc.newCondition.DeepCopy()
+	conditions := []*v1.PodCondition{}
```
```suggestion
	conditions := make([]*v1.PodCondition, 0, len(psuc.newConditions))
```

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -172,12 +202,17 @@ func conditionNeedsUpdate(status *v1.PodStatus, condition *v1.PodCondition) bool
-	if psuc.newCondition == nil {
+	if psuc.newConditions == nil {
 		return true
 	}
```
This can be removed now

KunWuLuan replied:
done

KunWuLuan committed:
refactor: update PodStatusPatchCall to support multiple conditions

- Change newCondition field to newConditions (slice of PodCondition pointers) to support multiple conditions
- Update syncStatus function signature to accept a slice of PodCondition instead of a single condition

This change enables the PodStatusPatchCall to handle multiple pod conditions in a single API call, improving efficiency and reducing the number of API calls needed when multiple conditions need to be updated simultaneously.

Signed-off-by: KunWuLuan <kunwuluan@gmail.com>

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -88,18 +92,21 @@ func (psuc *PodStatusPatchCall) Execute(ctx context.Context, client clientset.In
 		logger.V(3).Info("Updating pod condition", "pod", psuc.podRef, "conditionType", condition.Type, "conditionStatus", condition.Status, "conditionReason", condition.Reason)
 	}
 
+	anySynced := syncStatus(podStatusCopy, conditions, psuc.nominatingInfo)
```
Keep it below the comment about syncing

KunWuLuan replied:
done

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -147,9 +157,17 @@ func (psuc *PodStatusPatchCall) Merge(oldCall fwk.APICall) error {
+		for _, newCond := range psuc.newConditions {
+			if newCond.Type == cond.Type {
+				found = true
+				break
```
Worth having a test for such scenario

KunWuLuan replied:
Old condition should be remained, I understood.

KunWuLuan committed:
fix: fix condition update and merge logic in PodStatusPatchCall

Fixed several critical issues in PodStatusPatchCall:

1. Fixed logical error in syncStatus function:
   - Properly track condition update status to ensure return true when any condition is updated

2. Corrected behavior of PodStatusPatchCall.Merge method:
   - Modified merge logic to prioritize conditions from old calls over new calls
   - Ensure old call conditions are properly preserved during merge

3. Added comprehensive test cases:
   - Added unit tests for syncStatus function
   - Added test cases to verify condition update behavior in different scenarios
   - Added tests to verify correct merge logic

These changes ensure accurate Pod status updates and correct priority during API call merging.

Fixes logical errors to ensure Pod conditions and nomination info are correctly synced to the API server.

Signed-off-by: KunWuLuan <kunwuluan@gmail.com>

KunWuLuan commented on pkg/scheduler/schedule_one.go:
```diff
@@ -1142,7 +1142,11 @@ func truncateMessage(message string) string {
+		conditions := make([]*v1.PodCondition, 0)
+		if condition != nil {
+			conditions = append(conditions, condition)
+		}
```
done

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -69,10 +69,17 @@ func (psuc *PodStatusPatchCall) UID() types.UID {
 	nnnNeedsUpdate := nominatingInfo.Mode() == fwk.ModeOverride && status.NominatedNodeName != nominatingInfo.NominatedNodeName
-	if condition != nil {
-		if !podutil.UpdatePodCondition(status, condition) && !nnnNeedsUpdate {
+	if len(condition) > 0 {
```
You don't need to wrap this code into `if len(condition) > 0`

KunWuLuan committed:
Update pkg/scheduler/framework/api_calls/pod_status_patch.go

Co-authored-by: Maciej Skoczeń <87243939+macsko@users.noreply.github.com>

KunWuLuan committed:
Update pkg/scheduler/framework/api_calls/pod_status_patch.go

Co-authored-by: Maciej Skoczeń <87243939+macsko@users.noreply.github.com>

KunWuLuan committed:
1. remove the condition length check in pkg/scheduler/framework/api_calls/pod_status_patch.go
2. overwrite the condition of a old call in *PodStatusPatchCall.Merge

Signed-off-by: KunWuLuan <kunwuluan@gmail.com>

KunWuLuan commented:
/remove-lifecycle stale

KunWuLuan commented:
@macsko @helayoty Gentle ping on this one. Would love to get your thoughts when you have time. Thanks!

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -120,19 +126,22 @@ func (psuc *PodStatusPatchCall) Sync(obj metav1.Object) (metav1.Object, error) {
 		psuc.podStatus = pod.Status.DeepCopy()
 	}
-	newCondition := psuc.newCondition.DeepCopy()
+	for _, condition := range psuc.newConditions {
```
Initialize the `newConditions` with the expected size just above the for loop

KunWuLuan replied:
Thanks for the review. Fixed in commit 3f66e128.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch.go:
```diff
@@ -120,19 +126,22 @@ func (psuc *PodStatusPatchCall) Sync(obj metav1.Object) (metav1.Object, error) {
-	// Sync passed pod's status with the call's condition and nominatingInfo.
-	synced := syncStatus(&podCopy.Status, newCondition, psuc.nominatingInfo)
-	if !synced {
+	// Sync status to have the condition and nominatingInfo applied on a podStatusCopy.
```
There is no `podStatusCopy`. Could you update the comment to better reflect the code?

KunWuLuan replied:
Thanks for the review. Fixed in commit 3f66e128.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -47,49 +47,49 @@ func TestPodStatusPatchCall_IsNoOp(t *testing.T) {
 			name:           "No-op when condition and node name match",
 			pod:            podWithNode,
-			condition:      &v1.PodCondition{Type: v1.PodScheduled, Status: v1.ConditionFalse},
+			conditions:     []*v1.PodCondition{{Type: v1.PodScheduled, Status: v1.ConditionFalse}},
```
Add a test case with multiple conditions

KunWuLuan replied:
Fixed in commit 3f66e12

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -132,8 +132,23 @@ func TestPodStatusPatchCall_Merge(t *testing.T) {
+		}
+		if newCall.newConditions[1].Type != v1.PodInitialized || newCall.newConditions[1].Status != v1.ConditionTrue {
+			t.Errorf("Expected PodScheduled condition, but got: %v", newCall.newConditions)
 		}
```
```suggestion
		expectedConditions := []*v1.PodCondition{{Type: v1.PodScheduled, Status: v1.ConditionTrue}, {Type: v1.PodInitialized, Status: v1.ConditionTrue}}
		if diff := cmp.Diff(expectedConditions, newCall.newConditions); diff != "" {
			t.Errorf("Unexpected conditions after merging (-want,+got)\n%s", diff)
		}
```

KunWuLuan replied:
Fixed in the latest commit.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -196,6 +211,148 @@ func TestPodStatusPatchCall_Sync(t *testing.T) {
+func TestSyncStatus(t *testing.T) {
+	tests := []struct {
+		name                    string
+		podStatusBefore         *v1.PodStatus
```
Maybe?
```suggestion
		oldPodStatus         *v1.PodStatus
```

KunWuLuan replied:
Fixed in the latest commit.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -196,6 +211,148 @@ func TestPodStatusPatchCall_Sync(t *testing.T) {
+		podStatusBefore         *v1.PodStatus
+		conditions              []*v1.PodCondition
+		nominatingInfo          *fwk.NominatingInfo
+		expectedResult          bool
```
```suggestion
		expectUpdated          bool
```

KunWuLuan replied:
Fixed in the latest commit.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -196,6 +211,148 @@ func TestPodStatusPatchCall_Sync(t *testing.T) {
+		nominatingInfo          *fwk.NominatingInfo
+		expectedResult          bool
+		expectedNominatedNode   string
+		expectedConditionStatus map[v1.PodConditionType]v1.ConditionStatus
```
Why not just expect conditions slice and compare it using cmp.Diff?

KunWuLuan replied:
Fixed in the latest commit.

macsko commented on pkg/scheduler/framework/api_calls/pod_status_patch_test.go:
```diff
@@ -132,8 +132,23 @@ func TestPodStatusPatchCall_Merge(t *testing.T) {
+		}
+	})
+
+	t.Run("Overwrite nominating info and condition of a old call", func(t *testing.T) {
```
It would be also good to have a test where the oldCall has a condition that newCall doesn't have

KunWuLuan replied:
Fixed in the latest commit.

macsko commented:
This also requires a release note as it modifies scheduler's interface

KunWuLuan committed:
Address review comments for multi-condition PodStatusPatch support

- Initialize newConditions with expected capacity in Sync method
- Fix comment referencing podStatusCopy instead of podCopy
- Use cmp.Diff with cmpopts.IgnoreFields for condition comparisons in tests
- Rename test fields: podStatusBefore -> oldPodStatus, expectedResult -> expectUpdated
- Replace expectedConditionStatus map with expectedConditions slice in TestSyncStatus
- Add test case for IsNoOp with multiple differing conditions
- Add test case for Merge when old call has a condition new call lacks

Signed-off-by: KunWuLuan <kunwuluan@gmail.com>

KunWuLuan commented:
> This also requires a release note as it modifies scheduler's interface.

Updated the PR description, thanks!

macsko commented:
/lgtm
/approve

pacoxu commented:
/kind feature
